src/main/kotlin/org/jitsi/jicofo/conference/source/ValidatingConferenceSourceMap.kt

```diff
@@ -18,67 +18,302 @@
+            }
+            source.msid?.let { msid ->
+                if (msidToOwnerMap.containsKey(msid) && msidToOwnerMap[msid] != owner) {
+                    throw MsidConflictException(msid)
+                }
+            }
+            if (existingSourceSet.sources.size + sourcesToAdd.sources.size > maxSsrcsPerUser) {
+                throw SsrcLimitExceededException(maxSsrcsPerUser)
+            }
+        }
+
+        // If we rejected any of the SSRCs we'd throw above. We accepted them all.
+        val resultingSources = existingSourceSet.sources + sourcesToAdd.sources
+        val resultingSsrcs = resultingSources.map { it.ssrc }.toSet()
 
-        val (sourcesToRemove2, groupsToRemove) = ConferenceSourceMap(owner, sourcesToAdd).toMediaSourceMap()
-        try {
-            val added = validator.tryAddSourcesAndGroups(sourcesToRemove2, groupsToRemove)
-            val addedConferenceSourceMap = fromMediaSourceMap(
-                added[0] as MediaSourceMap,
-                added[1] as MediaSourceGroupMap
-            )
-            add(addedConferenceSourceMap)
-            return addedConferenceSourceMap
-        } catch (e: InvalidSSRCsException) {
-            throw ValidationFailedException(e.message)
+        val acceptedGroups = mutableSetOf<SsrcGroup>()
+        sourcesToAdd.ssrcGroups.forEach {
+            // We just ignore empty groups and groups which have been signaled multiple times.
+            when {
+                it.ssrcs.isEmpty() -> logger.info("Empty group signaled, ignoring.")
+                existingSourceSet.ssrcGroups.contains(it) -> logger.info("Duplicate group signaled, ignoring.")
+                !resultingSsrcs.containsAll(it.ssrcs) -> throw GroupContainsUnknownSourceException(it.ssrcs)
+                else -> acceptedGroups.add(it)
+            }
         }
+
+        val resultingSourceSet = EndpointSourceSet(resultingSources, existingSourceSet.ssrcGroups + acceptedGroups)
+        validateEndpointSourceSet(resultingSourceSet)
+
+        val acceptedSourceMap = ConferenceSourceMap(
+            owner to EndpointSourceSet(existingSourceSet.sources + sourcesToAdd.sources, acceptedGroups)
+        )
+        add(acceptedSourceMap)
+        return acceptedSourceMap
     }
 
+    /**
+     * Attempts to remove [sourcesToRemove] as sources owned by [owner]. The attempt is successful if all referenced
+     * sources and groups are indeed owned by [owner], and the resulting set of sources for [owner] is valid (see
+     * [validateEndpointSourceSet]).
+     *
+     * If the attempt is successful, the sources are removed from [owner]'s sources and the set of removed sources is
+     * returned. The set of removed sources might differ from [sourcesToRemove] because:
+     * 1. We only match a source's SSRC, so a source added with parameters (MSID, etc) can be removed by simply
+     * referencing the SSRC.
+     * 2. We automatically remove any groups that contain any of the removed sources.
```

Comment:
Should this be "that contain all of the removed sources"?  What do we do if one source in a group is removed?  Or will that cause a validation failure?

Reply:
We require that all groups' SSRCs be described as sources. If one source in a group is removed, it may result in a validation failure or not, depending on what remains. 

For example, if we remove source 1 from this we end up with a valid set (the group is removed, only source 2 remains):
```
sources = 1, 2
groups = sim(1,2)
```

But if we remove source 1 from this, we end up with an invalid set (sources 2 and 3 have the same MSID but are not grouped in any way):
```
sources = 1, 2, 3
groups = sim(1, 2, 3)
```

I think this is new. Previously we didn't automatically remove groups. If you think it may be confusing or lead to problems I can remove it.

Reply:
Yeah, I'm wondering what would happen if an endpoint decided to go from a three-source SIM group to a two-source SIM group.  Do we even have a way to signal that now?

Reply:
Hm, no. It would have to do a `remove` followed by an `add`.

Reply:
That should be fine then, I guess.